rust_dev_preview/examples/auto-scaling/src/scenario.rs

```diff
@@ -0,0 +1,675 @@
+    group: Result<Vec<String>, ScenarioError>,
+    instances: Result<Vec<String>, anyhow::Error>,
+    activities: Result<Vec<Activity>, anyhow::Error>,
+}
+
+impl Display for AutoScalingScenarioDescription {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        writeln!(f, "\t      Group status:")?;
+        match &self.group {
+            Ok(groups) => {
+                for status in groups {
+                    writeln!(f, "\t\t- {status}")?;
+                }
+            }
+            Err(e) => writeln!(f, "\t\t! - {e}")?,
+        }
+        writeln!(f, "\t         Instances:")?;
+        match &self.instances {
+            Ok(instances) => {
+                for instance in instances {
+                    writeln!(f, "\t\t- {instance}")?;
+                }
+            }
+            Err(e) => writeln!(f, "\t\t! {e}")?,
+        }
+
+        writeln!(f, "\t        Activities:")?;
+        match &self.activities {
+            Ok(activities) => {
+                for activity in activities {
+                    writeln!(
+                        f,
+                        "\t\t- {} Progress: {}% Status: {:?} End: {:?}",
+                        activity.cause().unwrap_or("(Cause unknown)"),
+                        activity.progress(),
+                        activity.status_code(),
+                        // activity.status_message().unwrap_or_default()
+                        activity.end_time(),
+                    )?;
+                }
+            }
+            Err(e) => writeln!(f, "\t\t! {e}")?,
+        }
+
+        Ok(())
+    }
+}
+
+#[derive(Debug)]
+struct MetadataError {
+    message: Option<String>,
+    code: Option<String>,
+}
+
+impl MetadataError {
+    fn from(err: &dyn ProvideErrorMetadata) -> Self {
+        MetadataError {
+            message: err.message().map(|s| s.to_string()),
+            code: err.code().map(|s| s.to_string()),
+        }
+    }
+}
+
+impl Display for MetadataError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let display = match (&self.message, &self.code) {
+            (None, None) => "Unknown".to_string(),
+            (None, Some(code)) => format!("({code})"),
+            (Some(message), None) => message.to_string(),
+            (Some(message), Some(code)) => format!("{message} ({code})"),
+        };
+        write!(f, "{display}")
+    }
+}
+
+#[derive(Debug)]
+pub struct ScenarioError {
+    message: String,
+    context: Option<MetadataError>,
+}
+
+impl ScenarioError {
+    pub fn with(message: impl Into<String>) -> Self {
+        ScenarioError {
+            message: message.into(),
+            context: None,
+        }
+    }
+
+    pub fn new(message: impl Into<String>, err: &dyn ProvideErrorMetadata) -> Self {
+        ScenarioError {
+            message: message.into(),
+            context: Some(MetadataError::from(err)),
+        }
+    }
+}
+
+impl Error for ScenarioError {}
```

Comment:
I guess not setting `Source` is right here since we print the message in display?

Reply:
In context, an error looks like this:

```
Error: Failed to initialize scenario: Failed to find AZs: Request has expired. (RequestExpired)
```

Reply:
Done.